crates/codegen/ebnf/src/serializer.rs

```diff
@@ -0,0 +1,192 @@
+            }
+
+            if let Some(comment) = trailing_comment {
+                self.serialize_punctuation(" ");
+                self.serialize_comment(comment);
+            }
+
+            self.writer.end_line();
+        }
+    }
+
+    fn serialize_expr(&mut self, parent: &Expression) {
+        match parent {
+            Expression::Sequence { expressions } => {
+                self.serialize_child_expr(parent, &expressions[0]);
+
+                for expression in expressions.iter().skip(1) {
+                    self.serialize_punctuation(" ");
+                    self.serialize_child_expr(parent, expression);
+                }
+            }
+            Expression::Choice { expressions } => {
+                self.serialize_child_expr(parent, &expressions[0]);
+
+                for expression in expressions.iter().skip(1) {
+                    self.serialize_punctuation(" | ");
+                    self.serialize_child_expr(parent, expression);
+                }
+            }
+            Expression::Optional { expression } => {
+                self.serialize_child_expr(parent, expression);
+                self.serialize_punctuation("?");
+            }
+            Expression::ZeroOrMore { expression } => {
+                self.serialize_child_expr(parent, expression);
+                self.serialize_punctuation("*");
+            }
+            Expression::OneOrMore { expression } => {
+                self.serialize_child_expr(parent, expression);
+                self.serialize_punctuation("+");
+            }
+            Expression::Not { expression } => {
+                self.serialize_punctuation("!");
+                self.serialize_child_expr(parent, expression);
+            }
+            Expression::Range {
+                inclusive_start,
+                inclusive_end,
+            } => {
+                self.serialize_child_expr(parent, inclusive_start);
+                self.serialize_punctuation("…");
+                self.serialize_child_expr(parent, inclusive_end);
+            }
+            Expression::Atom { atom } => {
+                self.serialize_string_literal(atom);
+            }
+            Expression::Reference { reference } => {
+                self.serialize_identifier(reference);
+            }
+        };
+    }
+
+    fn serialize_child_expr(&mut self, parent: &Expression, child: &Expression) {
+        if discriminant(parent) != discriminant(child) && child.precedence() <= parent.precedence()
+        {
+            self.serialize_punctuation("(");
+            self.serialize_expr(child);
+            self.serialize_punctuation(")");
+        } else {
+            self.serialize_expr(child);
+        }
+    }
+
+    fn serialize_comment(&mut self, value: &String) {
+        self.writer.write_comment(format!("(* {value} *)"));
+    }
+
+    fn serialize_identifier(&mut self, name: &str) {
+        let value = &self.model.entries[name].ebnf_id;
```

Comment:
We don't emit the same panic message as above, in `fn serialize`. Is this intended?

Reply:
The above took its input from external callers, so more likely to fail. But it makes sense to show a message in both. Will modify.

Reply:
This makes sense and so we didn't need to change that; just wanted to ask to know if I'm missing something. Sorry for the churn!